### PR TITLE
crypto: replace rand<T>()%N idiom with unbiased rand_idx(N)

### PR DESCRIPTION
**Pull request by stoffu**

### CODE REVIEW

File: src/crypto/crypto.h [hunk 1]
```diff
@@ -162,6 +163,20 @@ namespace crypto {
     return res;
   }
 
+  /* Generate a random value between 0 and range_max-1 in an unbiased way
+   */
+  template<typename T>
+  typename std::enable_if<std::is_pod<T>::value, T>::type rand_range(T range_max) {
```
Comment (vtnerd) [inline src/crypto/crypto.h]: `is_pod` is not sufficient, `is_integral` is a bit better.

File: src/crypto/crypto.h [hunk 2]
```diff
@@ -162,6 +163,20 @@ namespace crypto {
...
+    typename std::remove_cv<T>::type res;
+    const typename std::remove_cv<T>::type max_value = (std::numeric_limits<T>::max() / range_max) * range_max;
+    do
+    {
+      res = rand<T>();
+    }
+    while (res >= max_value);
```
Comment (vtnerd) [inline src/crypto/crypto.h]: Is the lower bound of the range implied 0 or implied `std::numeric_limits<T>::min()` ? This implementation is currently the former, but I'm not sure if allowing signed types is a good thing, it seems like the implementation could bust in those situations.

**Comment by stoffu** [comment]
@vtnerd 
I've replaced `is_pod` with `is_unsigned` to limit the usage to the very common idiom (`rand()%N)` addressed here. If a need for handling more general cases arises later, a new function can then be added.

**Comment by vtnerd** [comment]
Sounds like I was unable to persuade you to use the STL function. Humbug. Might be worth handling the 0 case too. Or perhaps the documentation implies 0 should be filtered by caller?

**Comment by stoffu** [comment]
Now I'm persuaded to use the STL :)

File: src/common/dns_utils.cpp [hunk 3]
```diff
@@ -517,10 +517,7 @@ bool load_txt_records_from_dns(std::vector<std::string> &good_records, const std
   std::vector<std::vector<std::string> > records;
   records.resize(dns_urls.size());
 
-  std::random_device rd;
-  std::mt19937 gen(rd());
-  std::uniform_int_distribution<int> dis(0, dns_urls.size() - 1);
-  size_t first_index = dis(gen);
+  size_t first_index = crypto::rand_range<size_t>(0, dns_urls.size() - 1);
```
Comment (vtnerd) [inline src/common/dns_utils.cpp]: Well, getting to update this at the same time is nice.

File: src/crypto/crypto.h [hunk 4]
```diff
@@ -162,6 +163,16 @@ namespace crypto {
     return res;
   }
 
+  /* Generate a random value between range_min and range_max
+   */
+  template<typename T>
+  typename std::enable_if<std::is_integral<T>::value, T>::type rand_range(T range_min, T range_max) {
+    std::random_device rd;
```
Comment (vtnerd) [inline src/crypto/crypto.h]: We might want to define our own random generator to keep the behavior the same. The `mt19937` isn't a crypto-secure generator:
```c++
  struct random_device
  {
    typedef uint64_t result_type;
    static constexpr result_type min() { return 0; }
    static constexpr result_type max() { return result_type(-1); }
    result_type operator()() const { return crypto::rand<result_type>(); }
  };
```
will work directly with `uniform_int_distribution` and doesn't do a syscall each time (similar to `mt133937` usage).

**Comment by stoffu** [comment]
Thank you so much for the suggestion, I think the patch is a lot better than original now :)

**Comment by moneromooo-monero** [comment]
I liked the original version with a single param. Maybe it can be added back as a simple forwarder to the (0, N-1) call ?

**Comment by stoffu** [comment]
@moneromooo-monero 
I've added such a forwarder called `rand_idx`.

**Pull request merged**